### PR TITLE
Centering and scaling with sample-weighted train set

### PR DESCRIPTION
**Pull request by bhelfrecht**
I have tweaked the scalers/centerers to take account for sample weighting in the preprocessing, for example, to center by a weighted mean of the training samples.

### CODE REVIEW

**Review by Luthaf — COMMENTED** [review]
The code looks good, however I would love to see more documentation on the `sample_weight` parameter: what is it intended to do? How should I use it? Maybe you could extend the description a bit in the docstring; and add an example using it?

**Comment by Luthaf** [comment]
To be clear, I mostly mean taking the PR description and adding it to the code as a use case for this parameter 😃. Maybe also adding how this parameter is used.

File: tests/test_standard_flexible_scaler.py [hunk 1]
```diff
@@ -6,6 +6,16 @@
 
 
 class ScalerTests(unittest.TestCase):
+    def test_sample_weights(self):
```
Comment (Luthaf) [inline tests/test_standard_flexible_scaler.py:9]: Could you also add a test with _invalid_ `sample_weights`? What happen if the array shape is wrong? If the type of the parameter is wrong (`sample_weights=False`)?

**Review by Luthaf — APPROVED** [review]
Looks good to me! @rosecers I remember the procedure is to have a second quick review once the first round is done, would you like to do it yourself or nominate someone else?

**Comment by rosecers** [comment]
> Looks good to me! @rosecers I remember the procedure is to have a second quick review once the first round is done, would you like to do it yourself or nominate someone else?

I can do a second review -- diving in now

File: tests/test_kernel_flexible_centerer.py [hunk 2]
```diff
@@ -5,6 +5,36 @@
...
+        K_uniform_weighted = weighted_model.fit_transform(K, sample_weight=uniform_wts)
+        self.assertTrue(
+            (np.isclose(K_unweighted, K_uniform_weighted, atol=1e-12)).all()
+        )
+        K_nonuniform_weighted = weighted_model.fit_transform(
+            K, sample_weight=nonuniform_wts
+        )
+        self.assertFalse(
+            (np.isclose(K_unweighted, K_nonuniform_weighted, atol=1e-12)).all()
+        )
```
Comment (rosecers) [inline tests/test_kernel_flexible_centerer.py:25]: Superficial comment here -- can you name this something other than `nonuniform`? It's a bit jarring to have it named `nonuniform` and then `np.random.uniform`. I would more suggest:
```suggestion
        equal_wts = np.ones(len(K))
        unequal_wts = np.random.uniform(0, 100, size=(len(K),))
        
        model = KernelFlexibleCenterer()
        weighted_model = KernelFlexibleCenterer()
        
        K_wo_weights = model.fit_transform(K)
        K_equally_weighted = weighted_model.fit_transform(K, sample_weight=equal_wts)
        self.assertTrue(
            (np.isclose(K_wo_weights, K_equally_weighted, atol=1e-12)).all()
        )
        K_randomly_weighted = weighted_model.fit_transform(
            K, sample_weight=random_wts
        )
        self.assertFalse(
            (np.isclose(K_wo_weights, K_randomly_weighted, atol=1e-12)).all()
        )
```

File: skcosmo/preprocessing/flexible_scaler.py [hunk 3]
```diff
@@ -25,25 +25,33 @@ def __init__(self, with_mean=True, with_std=True, column_wise=False, tol=1e-15):
...
+        :type sample_weight: array of shape (n_samples,)
 
         :return: itself
         """
 
         self.n_samples_seen_, self.n_features_ = X.shape
+
+        if sample_weight is not None:
+            sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
+            sample_weight = sample_weight / np.sum(sample_weight)
```
Comment (rosecers) [inline skcosmo/preprocessing/flexible_scaler.py:44]: Docstring is confusing -- it can read that weights _should_ be normalized before passing. What happens if someone wants them non-normalized? Should we allow that?

File: skcosmo/preprocessing/flexible_scaler.py [hunk 4]
```diff
@@ -113,19 +124,28 @@ def __init__(self):
         """Initialize KernelFlexibleCenterer."""
         pass
 
-    def fit(self, K=None, y=None, K_fit_rows=None, K_fit_all=None):
+    def fit(self, K=None, y=None, sample_weight=None, K_fit_rows=None, K_fit_all=None):
```
Comment (rosecers) [inline skcosmo/preprocessing/flexible_scaler.py:127]: Does this fit with the sklearn-style syntax for KernelCenterer?

File: skcosmo/preprocessing/flexible_scaler.py [hunk 5]
```diff
@@ -113,19 +124,28 @@ def __init__(self):
...
         """Fit KernelFlexibleCenterer
 
         :param K: Kernel matrix
         :type K: ndarray of shape (n_samples, n_samples)
         :param y: ignored
+        :param sample_weight: weights for each sample. Sample weighting can be used to center (and scale) data using a weighted mean. Weights are normalized before preprocessing.
```
Comment (rosecers) [inline skcosmo/preprocessing/flexible_scaler.py:133]: same comment as above.

File: skcosmo/preprocessing/flexible_scaler.py [hunk 6]
```diff
@@ -195,12 +221,23 @@ def transform(self, K, y=None):
...
+        K_fit_rows=None,
+        K_fit_all=None,
+        **fit_params
+    ):
         r"""Fit to data, then transform it.
 
         :param K: Kernel matrix
         :type K: ndarray of shape (n_samples, n_samples)
         :param y: ignored
+        :param sample_weight: weights for each sample. Sample weighting can be used to center (and scale) data using a weighted mean. Weights are normalized before preprocessing.
```
Comment (rosecers) [inline skcosmo/preprocessing/flexible_scaler.py:238]: ^^

File: skcosmo/preprocessing/flexible_scaler.py [hunk 7]
```diff
@@ -240,6 +283,9 @@ def fit(self, Knm, Kmm, y=None):
         :type Kmm: ndarray of shape (n_active, n_active)
 
         :param y: ignored
+        :param sample_weight: weights for each sample. Sample weighting can be used to center (and scale) data using a weighted mean. Weights are normalized before preprocessing.
```
Comment (rosecers) [inline skcosmo/preprocessing/flexible_scaler.py:286]: ^^

File: skcosmo/preprocessing/flexible_scaler.py [hunk 8]
```diff
@@ -73,17 +81,20 @@ def transform(self, X, y=None):
...
         return (X - self.mean_) / self.scale_
 
-    def fit_transform(self, X, y=None, **fit_params):
+    def fit_transform(self, X, y=None, sample_weight=None, **fit_params):
         r"""Fit to data, then transform it.
 
         :param X: Matrix
         :type X: ndarray
         :param y: ignored
+        :param sample_weight: weights for each sample. Sample weighting can be used to center (and scale) data using a weighted mean. Weights are normalized before preprocessing.
```
Comment (rosecers) [inline skcosmo/preprocessing/flexible_scaler.py:90]: same comment as above

File: skcosmo/preprocessing/flexible_scaler.py [hunk 9]
```diff
@@ -300,11 +350,14 @@ def fit_transform(self, Knm, Kmm, y=None, **fit_params):
         :type Kmm: ndarray of shape (n_active, n_active)
 
         :param y: ignored
+        :param sample_weight: weights for each sample. Sample weighting can be used to center (and scale) data using a weighted mean. Weights are normalized before preprocessing.
```
Comment (rosecers) [inline skcosmo/preprocessing/flexible_scaler.py:353]: same comment as above

File: tests/test_standard_flexible_scaler.py [hunk 10]
```diff
@@ -6,6 +6,36 @@
...
+        X_uniform_weighted = weighted_model.fit_transform(X, sample_weight=uniform_wts)
+        self.assertTrue(
+            (np.isclose(X_unweighted, X_uniform_weighted, atol=1e-12)).all()
+        )
+        X_nonuniform_weighted = weighted_model.fit_transform(
+            X, sample_weight=nonuniform_wts
+        )
+        self.assertFalse(
+            (np.isclose(X_unweighted, X_nonuniform_weighted, atol=1e-12)).all()
+        )
```
Comment (rosecers) [inline tests/test_standard_flexible_scaler.py:26]: same superficial naming comments as above

**Review by rosecers — COMMENTED** [review]
For the most part looks good, just a few clarifying comments

File: skcosmo/preprocessing/flexible_scaler.py (hunk 3 above)
Comment (bhelfrecht) [reply skcosmo/preprocessing/flexible_scaler.py:44]: I have tweaked the docstrings slightly to clarify that the weights are normalized automatically. I think the weights should definitely be normalized. For instance, if you didn't normalize the weights and assigned each sample a weight of 2, you would end up centering by 2x the mean.

File: skcosmo/preprocessing/flexible_scaler.py (hunk 4 above)
Comment (bhelfrecht) [reply skcosmo/preprocessing/flexible_scaler.py:127]: sklearn's `KernelCenterer` doesn't allow for sample weights, but the `sample_weight` argument is consistent with models that do allow it in the `fit` call, e.g., `Ridge`. The `fit` call for sklearn's `KernelCenterer` also doesn't take `K_fit_rows` or `K_fit_all`, but those were there already.

File: tests/test_kernel_flexible_centerer.py (hunk 2 above)
Comment (bhelfrecht) [reply tests/test_kernel_flexible_centerer.py]: I have changed all of the "uniform" to "equal" and "nonuniform" to "nonequal"

**Comment by bhelfrecht** [comment]
A comment on the `KernelNormalizer`: I have removed the `fit` options for passing `K_fit_rows` and `K_fit_all`:

1. The only reason I can see for passing these is if you wanted to do a "fake fit", i.e., copy the centerer if your training data is so massive that you can't even afford three (cheap) mean calls, and there are already options for this that don't clutter the `fit` call. For instance, you can (a) just copy/pickle/JSON serialize a previously fitted model, or (b) set `K_fit_rows_` and `K_fit_all_` by creating an instance of `KernelNormalizer` and then adding these to its `__dict__`. The fit required `K` regardless of whether `K_fit_rows` and `K_fit_all` were passed, which kind of defeats the purpose of doing the "fake fit" in this way anyway.
2. It really makes handling sample weighting messier than it already is, but this is minor compared to (1)

**Review by rosecers — APPROVED** [review]
